CRM/Core/Form.php

```diff
@@ -1109,6 +1119,141 @@ public function addSelect($name, $props = array(), $required = FALSE) {
+    if (!isset($props['context'])) {
+      $props['context'] = $this->getDefaultContext();
+    }
+    // Resolve entity.
+    if (!isset($props['entity'])) {
+      $props['entity'] = $this->getDefaultEntity();
+    }
+    // Resolve field.
+    if (!isset($props['field'])) {
+      $props['field'] = strrpos($name, '[') ? rtrim(substr($name, 1 + strrpos($name, '[')), ']') : $name;
+    }
+    // Resolve fields properties from API.
+    $info = civicrm_api3($props['entity'], 'getfields');
+    foreach ($info['values'] as $uniqueName => $fieldSpec) {
+      if (
+          $uniqueName === $props['field'] ||
+          CRM_Utils_Array::value('name', $fieldSpec) === $props['field'] ||
+          in_array($props['field'], CRM_Utils_Array::value('api.aliases', $fieldSpec, array()))
+      ) {
+        break;
+      }
+    }
+
+    $label = isset($props['label']) ? $props['label'] : $fieldSpec['title'];
+
+    $widget = isset($props['type']) ? $props['type'] : $fieldSpec['html']['type'];
+    if ($widget == 'TextArea' && $props['context'] == 'search') {
+      $widget = 'Text';
+    }
+
+    $isSelect = (in_array($widget, array(
+          'Select',
+          'Multi-Select',
+          'Select State/Province',
+          'Multi-Select State/Province',
+          'Select Country',
+          'Multi-Select Country',
+          'AdvMulti-Select',
+          'CheckBox',
+          'Radio',
+    )));
+
+    if ($isSelect) {
+      // Fetch options from the api unless passed explicitly.
+      if (isset($props['options'])) {
+        $options = $props['options'];
+      }
+      else {
+        $info = civicrm_api3($props['entity'], 'getoptions', $props);
+        $options = $info['values'];
+      }
+
+      // The placeholder is only used for select-elements.
+      if (!array_key_exists('placeholder', $props)) {
+        $props['placeholder'] = $required ? ts('- select -') : $props['context'] == 'search' ? ts('- any -') : ts('- none -');
+      }
+
+      if ($props['context'] == 'search' || ($widget !== 'AdvMulti-Select' && strpos($widget, 'Select') !== FALSE)) {
+        $widget = 'Select';
+      }
+      $props['class'] = (isset($props['class']) ? $props['class'] . ' ' : '') . "crm-select2";
+      if ($props['context'] == 'search' || strpos($widget, 'Multi') !== FALSE) {
+        $props['class'] .= ' huge';
+        $props['multiple'] = 'multiple';
+      }
+
+      // Add data for popup link.
+      if ($props['context'] != 'search' && $widget == 'Select' && CRM_Core_Permission::check('administer CiviCRM')) {
+        $props['data-option-edit-path'] = array_key_exists('option_url', $props) ? $props['option_url'] : $props['data-option-edit-path'] = CRM_Core_PseudoConstant::getOptionEditUrl($fieldSpec);
+        $props['data-api-entity'] = $props['entity'];
+        $props['data-api-field'] = $props['field'];
+      }
+    }
+
+    CRM_Utils_Array::remove($props, 'entity', 'field', 'context', 'label');
+    switch ($widget) {
```

Comment:
A sense of deja vu:
 * https://github.com/civicrm/civicrm-core/blob/master/CRM/Core/BAO/CustomField.php#L828
 * https://github.com/civicrm/civicrm-core/blob/master/CRM/Core/BAO/UFGroup.php#L1837

IMHO, these switch/if-else structures are significant to understanding Civi-Quckform's history/mistakes. The switch() statement is quite understandable as an intermediate step, but when it lingers in the codebase it becomes something scary that many developers avoid (prob due to a mix of TLDR-itis and conservatism... but maybe I'm projecting...). Anyway, it's pretty important that form components be represented in decoupled, bite-sized chunks (like QuickForm elements or jQuery plugins or Angular directives, ad nauseum).

With that said, maybe the simple act of moving the switch from CRM_Core_BAO_CustomGroup to CRM_Core_Form would make a major change in perspective...